### DISCUSSION TITLE
The whereHas condition is not being applied when using closure on relationship with existing whereHas

### DISCUSSION THREAD

**Discussion by jmrecodes** [laravel/framework#53196] [Q&A]
### Laravel Version

11.4

### PHP Version

8.3

### Description

The issue is that when using a closure with the `whereHas` method on a relationship that already contains a `whereHas` method, the additional condition is not being included in the generated SQL query:

{
  "sql_without_closure": "select * from \"invoice_product_relationships\" where \"invoice_product_relationships\".\"invoice_id\" = ? and \"invoice_product_relationships\".\"invoice_id\" is not null and exists (select * from \"products\" where \"invoice_product_relationships\".\"product_id\" = \"products\".\"id\" and \"deleted_at\" is null and \"products\".\"deleted_at\" is null) and exists (select * from \"products\" where \"invoice_product_relationships\".\"product_id\" = \"products\".\"id\" and \"is_membership_product\" = ? and \"products\".\"deleted_at\" is null)",

  "bindings_without_closure": [2, true],

  "sql_with_closure": "select * from \"invoice_product_relationships\" where \"invoice_product_relationships\".\"invoice_id\" = ? and \"invoice_product_relationships\".\"invoice_id\" is not null and exists (select * from \"products\" where \"invoice_product_relationships\".\"product_id\" = \"products\".\"id\" and \"deleted_at\" is null and \"products\".\"deleted_at\" is null)",

  "bindings_with_closure": [2]
}

Notes:

The sql_without_closure includes the 'is_membership_product' condition as expected.
bindings_without_closure has two bindings: one for invoice_id and one for is_membership_product.
BUG: The sql_with_closure does NOT include the 'is_membership_product' condition.
BUG: bindings_with_closure only includes one binding (for invoice_id), missing the is_membership_product binding.

### Steps To Reproduce

1. In the Laravel project, create the following migration files:

> php artisan make:migration create_invoices_table
> php artisan make:migration create_products_table
> php artisan make:migration create_invoice_product_relationships_table

2. Fill the migrations with the following content:

```
// create_invoices_table
public function up()
{
    Schema::create('invoices', function (Blueprint $table) {
        $table->id();
        $table->timestamps();
    });
}
```

```
// create_products_table
public function up()
{
    Schema::create('products', function (Blueprint $table) {
        $table->id();
        $table->boolean('is_membership_product')->default(false);
        $table->softDeletes();
        $table->timestamps();
    });
}
```

```
// create_invoice_product_relationships_table
public function up()
{
    Schema::create('invoice_product_relationships', function (Blueprint $table) {
        $table->id();
        $table->foreignId('invoice_id')->constrained();
        $table->foreignId('product_id')->constrained();
        $table->timestamps();
    });
}
```

3. Create the following models:

> php artisan make:model Invoice
> php artisan make:model Product
> php artisan make:model InvoiceProductRelationship

4. Fill the models with the following content:

```
// App\Models\Invoice
class Invoice extends Model
{
    public function lineItems()
    {
        return $this->hasMany(InvoiceProductRelationship::class)->whereHas('product', function ($query) {
            $query->whereNull('deleted_at');
        });
    }
}
```

```
// App\Models\Product
class Product extends Model
{
    use SoftDeletes;

    protected $fillable = ['is_membership_product'];
}
```

```
// App\Models\InvoiceProductRelationship
class InvoiceProductRelationship extends Model
{
    public function product()
    {
        return $this->belongsTo(Product::class);
    }
}
```

5. Create the BugDemoController with the following content:

```
<?php

namespace App\Http\Controllers;

use App\Models\Invoice;
use App\Models\Product;
use App\Models\InvoiceProductRelationship;

class BugDemoController extends Controller
{
    public function demonstrateBug()
    {
        // Create test data
        $invoice = Invoice::create();
        $product1 = Product::create(['is_membership_product' => true]);
        $product2 = Product::create(['is_membership_product' => false]);
        InvoiceProductRelationship::create(['invoice_id' => $invoice->id, 'product_id' => $product1->id]);
        InvoiceProductRelationship::create(['invoice_id' => $invoice->id, 'product_id' => $product2->id]);

        // Demonstrate the bug
        $queryWithoutClosure = $invoice->lineItems()->whereHas('product', function ($query) {
            $query->where('is_membership_product', true);
        });
        $queryWithClosure = $invoice->lineItems(function ($query) {
            $query->whereHas('product', function ($productQuery) {
                $productQuery->where('is_membership_product', true);
            });
        });

        return response()->json([
            'sql_without_closure' => $queryWithoutClosure->toSql(),
            'bindings_without_closure' => $queryWithoutClosure->getBindings(),
            'sql_with_closure' => $queryWithClosure->toSql(),
            'bindings_with_closure' => $queryWithClosure->getBindings(),
        ]);
    }
}
```

6. Add the route to routes/web.php:

```
use App\Http\Controllers\BugDemoController;

Route::get('/', [BugDemoController::class, 'demonstrateBug']);
```

7. Run the following commands:

> composer install
> php artisan migrate
> php artisan serve

8. Visit http://localhost:8000 in your browser and observe the output.

The demo's repo can be found at https://github.com/jmrecodes/laravel-whereHas-bug-demo

**Comment by staudenmeir** [comment]
Hi @jmrecodes,

Your `$queryWithClosure` query doesn't include the `is_membership_product` constraint because the closure doesn't get used. You pass it as an argument to `Invoice::lineItems()`, but the relationships doesn't have any parameters.

**Comment by macropay-solutions** [comment]
@staudenmeir We were about to ask if this is a new feature of Laravel 11, passing conditions as params that are not declared in the relation's definition.


```php
        $queryWithClosure = $invoice->lineItems(function ($query) {
            $query->whereHas('product', function ($productQuery) {
                $productQuery->where('is_membership_product', true);
            });
        });

    public function lineItems()
    {
        return $this->hasMany(InvoiceProductRelationship::class)->whereHas('product', function ($query) {
            $query->whereNull('deleted_at');
        });
    }

```
